### PR TITLE
working registration with flirt step, bias correct change necessary

### PR DESCRIPTION
**Pull request by damaggu**

### CODE REVIEW

File: samri/optimization/registration.py [hunk 1]
```diff
@@ -108,14 +109,21 @@ def structural_rigid_affine(template="/Users/marksm/GitHub/mriPipeline/ants_test
...
+	n4.inputs.bspline_fitting_distance = 10
+	# n4.inputs.shrink_factor = 2
 	n4.inputs.n_iterations = [200,200,200,200]
 	n4.inputs.convergence_threshold = 1e-11
 	# n4.inputs.output_image = 'ss_n4_{}_ofM{}.nii.gz'.format(participant,i)
-	n4.inputs.output_image = 'hard.nii.gz'
+	n4.inputs.output_image = 'hard_flirt.nii.gz'
 	n4_res = n4.run()
 
+	flt = fsl.FLIRT(cost_func='corratio', dof = 6, searchr_x = [-180,180], searchr_y = [-180,180], searchr_z = [-180,180], force_scaling = True)
```
Comment (TheChymera) [inline samri/optimization/registration.py:120]: are `searchr_x` and friends the parameters that make the registration try to reorient the image?

Comment (damaggu) [reply samri/optimization/registration.py:120]: yup

**Comment by TheChymera** [comment]
I'm glad this worked out for you, and I think it's ok to pull it for the registration examples - but I would be reticent to use this in the pipeline. We would either have to create two pipelines just to accommodate for this, or further parameterize the Bruker pipeline to allow selection of FLIRT or ANTs. I think this is too much complexity atm, and ideally we would figure out how to do this in ANTs.

I think the people in [this thread](https://github.com/stnava/ANTs/issues/399#issuecomment-272841079) would be happy to help us figure out how to use their software to accieve similar or better performance as FNIRT. Actually, I think  [this commnt](https://github.com/stnava/ANTs/issues/399#issuecomment-272841079) already points out some `antsRegistration` inputs that would help acchieve the same effect. Hve you tested them?

**Comment by damaggu** [comment]
Why not incorporate as boolean flag, which gets activated if by hand or automated in case source and template orientation are not equal? or maybe it works just fine before the ants registration? have you tried for your data?

The -r option is not a real solution, since it's a hand-written a priori transformation, hence the user would need to the transformation steps, in which case he/she could do it with fslorient/swapdim as well.

**Comment by TheChymera** [comment]
ah, I see. I thought there was a way to tell FNIRT to automatically go through all alternate registrations (it's just 8, I believe) and pick the best one. As such I feel the advantage this has over ANTs grows increasingly dim - this is basically what one could do with nibabel/(fslswapdim+fslorient) + ANTs

**Comment by damaggu** [comment]
no, the hardcoded part i was talking about was the -r option with antsRegistration. Flirt goes automatically through all angles

**Comment by TheChymera** [comment]
hm. uncool on the part of ANTs - so you think we can use FNIRT just to reorient the image? I really like ANTs and its interface, plus we finally managed to agree on it as a quasi-standard for the AIC.

But until ANTs gets this capability I really like the idea of your hand/automagically activated node - especially if we can keep all the transformations except this reorienting done in ANTs.

**Comment by TheChymera** [comment]
But it's a bit non-trivial to integrate complex logic into nipype. I still think we should ask the ANTs people a few more times before we go down that route.

**Comment by damaggu** [comment]
flirt could serve as initial rigid transform in front of ants, it works with 6 dof. no logic

File: samri/optimization/registration.py [hunk 2]
```diff
@@ -96,8 +97,8 @@ def structural_per_participant_test(participant,
 			res = struct_registration.run()
 
 def structural_rigid_affine(template="/Users/marksm/GitHub/mriPipeline/ants_test/template4.nii.gz",
```
Comment (TheChymera) [inline samri/optimization/registration.py:99]: Looking good - so this function here now uses FLIRT for the reorient, and ANTs for everything else? Perhaps we should edit the name accordingly? and add a short docstring stating as much?

File: samri/optimization/registration.py [hunk 3]
```diff
@@ -145,10 +153,73 @@ def structural_rigid_affine(template="/Users/marksm/GitHub/mriPipeline/ants_test
 	struct_registration.inputs.args = '--float'
 	struct_registration.inputs.num_threads = 6
 
-	struct_registration.inputs.moving_image = n4_res.outputs.output_image
+	struct_registration.inputs.moving_image = flt_res.outputs.out_file
+	# struct_registration.inputs.output_warped_image = 'ss_{}_ofM{}.nii.gz'.format(participant,i)
+	struct_registration.inputs.output_warped_image = output_image
+	res = struct_registration.run()
+
+def structural_flirt_affine(template="/Users/marksm/GitHub/mriPipeline/ants_test/template4.nii.gz",
```
Comment (TheChymera) [inline samri/optimization/registration.py]: and this function here uses FLIRT for all things linear? Again, maybe a more descriptive name and/or a docstring?

**Review by TheChymera — CHANGES_REQUESTED** [review]

File: samri/optimization/registration.py (hunk 2 above)
Comment (damaggu) [reply samri/optimization/registration.py:99]: I changed it accordingly. The function name is now structural_rigid_flirt_nonlin_syn, with added docstring. I've been playing with the parameters a little. Now it works really well for me. 

Could you try on your data?

**Comment by damaggu** [comment]
how does the function work for your files, did you test it?

**Comment by TheChymera** [comment]
not yet, as I do not have a whole-head mouse T2 atlas. I am really having trouble finding one. Could you maybe help me? Dis the resource where you found the rat atlas also have something for mice?

**Comment by damaggu** [comment]
would this one work? https://scalablebrainatlas.incf.org/mouse/WHS12 or any of these
https://scalablebrainatlas.incf.org/main/index.php?

**Comment by damaggu** [comment]
otherwise i found mine here: https://scalablebrainatlas.incf.org/main/index.php?

**Comment by TheChymera** [comment]
I haven't been able to access the original data in full resolutions  - I believe this is the link https://civmvoxport.duhs.duke.edu/voxbase/login.php?return_url=%2Fvoxbase%2Fstudyhome.php%3Fstudyid%3D132

But I did download one of the downsized templates, and it's not really that cool. Most of the interesting extra-cerebral features are cropped, and the template is not really regularly shaped/oriented:

screenshot from 2017-01-25 16-14-16


I'd really like to get ahold of the second atlas listed here - but I can't seem to get through to these researchers: https://wiki.mouseimaging.ca/display/MICePub/Mouse+Brain+Atlases

**Pull request merged**